deps/spidershim/src/v8string.cc

```diff
@@ -79,9 +79,20 @@ MaybeLocal<String> String::NewFromUtf8(Isolate* isolate, const char* data,
                                        v8::NewStringType type, int length) {
   assert(type == v8::NewStringType::kNormal); // TODO: Add support for interned strings
   JSContext* cx = JSContextFromIsolate(isolate);
-  JS::RootedString str(cx, length >= 0 ?
-                        JS_NewStringCopyN(cx, data, length) :
-                        JS_NewStringCopyZ(cx, data));
+
+  // In V8's api.cc, this conditional block is annotated with the comment:
+  //    TODO(dcarney): throw a context free exception.
+  if (length > v8::String::kMaxLength) {
+    return MaybeLocal<String>();
+  }
+
+  if (length < 0) {
+    length = strlen(data);
+  }
+
+  size_t twoByteLen;
+  char16_t* twoByteChars = JS::UTF8CharsToNewTwoByteCharsZ(cx, JS::UTF8Chars(data, length), &twoByteLen).get();
```

Comment:
Hmm, I might misunderstand, but I think this only leaks when the JS_NewUCString call below fails, since that function takes ownership of twoByteChars on success, per <https://dxr.mozilla.org/mozilla-central/rev/fc15477ce628599519cb0055f52cc195d640dc94/js/src/jsapi.h#4520-4524>.

If this declared twoByteChars as UniqueTwoByteChars, then it would have to use JS_NewUCStringCopyN to copy twoByteChars instead of reusing it, as on this branch:

https://github.com/mozilla/spidernode/compare/master...mykmelez:free-two-byte-chars-automatically

And that seems like an unnecessary extra allocation. So I would rather explicitly free twoByteChars on error, as in pull request #47.

How does that sound?